### PR TITLE
Add Register iterator

### LINKED ISSUE #136: Rust: implement FromStr for Register
Hey! Thanks for writing a fantastic disassembler.

I'd like to convert user-provided strings to register enums in a [CLI tool](https://github.com/entropic-security/xgadget) that uses the Rust library. Supporting this functionality might be useful for other purposes down the line, like #77.

AFAIK there are two ways to do this (both adding a optional dependencies gated by the `std` feature):
- The simplest is to use the [`strum` crate's `EnumString`](https://docs.rs/strum/0.20.0/strum/derive.EnumString.html) macro.
- I think a better fit here would be a combo of [`num-traits` crate's `FromPrimitive`](https://docs.rs/num-traits/0.2.14/num_traits/cast/trait.FromPrimitive.html) and using the existing `GEN_DEBUG_REGISTER` array. This approach gains both `String` -> `Register` and `u32`/`u64`/`usize`, etc -> `Register`.

I've tried out the latter in [this commit to my fork](https://github.com/tnballo/iced/commit/19ca3470e8ed4e2fd7abfb32ceffdb421b9e819f) (without updating the generator code, which I assume is a pre-req to PR-ing).

Please let me know what you think!

### PR DESCRIPTION
**Pull request by tnballo**
Implements preferred solution to #136.

Fixes #136

### CODE REVIEW

**Comment by tnballo** [comment]
Hmm...got 7 CI errors for files I did not modify in 8f3ce3f. Fixed those in 4a9a081, another 10 errors crop up - again for files unrelated to the change I want to make. Guessing this is due to a version increase for the Rust toolchain since the last time a PR was issued.

Should I fix all errors? They look uncontroversial but are outside the scope of this PR.

File: src/rust/iced-x86/src/register.rs [hunk 1]
```diff
@@ -1467,6 +1468,12 @@ impl Default for Register {
 // GENERATOR-END: Register
 
 impl Register {
+	/// Returns an iterator for `Register` enum values
+	#[must_use]
+	#[inline]
+	pub fn values() -> RegisterIterator {
+		RegisterIterator::new()
+	}
```
Comment (0xd4d) [inline src/rust/iced-x86/src/register.rs:1476]: Add a newline here

**Review by 0xd4d — CHANGES_REQUESTED** [review]

**Comment by 0xd4d** [comment]
Yes, undo that last commit (clippy fixes), I'll just test it locally.

Also move the register iterator to register.rs.

File: src/rust/iced-x86/src/register_iterator.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct RegisterIterator {
+	curr_val: u8,
+}
+
+impl RegisterIterator {
+	/// Creates a new iterator
+	#[must_use]
+	#[inline]
+	pub fn new() -> Self {
```
Comment (0xd4d) [inline src/rust/iced-x86/src/register_iterator.rs:17]: This doesn't need to be public. Users should always call `Register::values()` to get the iterator.

**Comment by 0xd4d** [comment]
I've pushed the clippy fixes to master so if you rebase, the CI should (hopefully) pass.

**Comment by tnballo** [comment]
Thanks! Reverted the clippy commit, re-based on master, and made the review changes. Only 2 CI errors this time (subset of the revert).

**Comment by tnballo** [comment]
Those 2 CI errors were my mistake with git workflows - PR is now a single file change (`register.rs`). Looks like Rust CI passed but Python had an issue downloading dependencies.

**Review by 0xd4d — APPROVED** [review]
Yeah it's a network error, I've seen it happen before:

"warning: spurious network error (2 tries remaining): error inflating zlib stream; class=Zlib (5)"

**Comment by 0xd4d** [comment]
Thanks, it's merged now!

BTW, the `Register` enum uses names `R8L` - `R15L` but most assemblers use the names `R8B`-`R15B`, so you might need to special case those regs in xgadget in case the users use `r8b`.

**Comment by tnballo** [comment]
Thanks for the heads up. With the hashmap logic you suggested in #136 it'll be quite convenient to add a 2nd key for those registers and then accept inputs case-insensitive :+1:

**Comment by wtfsck** [comment]
@tnballo Let me know if you need it pushed to crates.io and I'll release 1.10.3.

**Comment by tnballo** [comment]
Yup, if you could push to crates.io soon I'd appreciate it!

**Comment by wtfsck** [comment]
It's now on crates.io. I had to cherry-pick your commit and a few more since the master branch requires 1.42.0 but 1.10.x must support 1.20.0.

**Comment by tnballo** [comment]
Awesome, thanks!

**Pull request merged**